torchao/quantization/quantize_/workflows/float8/float8_tensor.py

```diff
@@ -0,0 +1,478 @@
+            _data,
+            scale,
+            block_size,
+            self.dtype,
+        )
+
+    def dequantize(self, output_dtype: Optional[torch.dtype] = None) -> torch.Tensor:
+        if output_dtype is None:
+            output_dtype = self.dtype
+
+        _data, scale = self._data, self.scale
+        return _dequantize_affine_float8(_data, scale, output_dtype)
+
+    @classmethod
+    def from_float(
+        cls,
+        hp_tensor: torch.Tensor,
+        dtype: torch.dtype,
+        granularity: FP8Granularity,
+    ):
+        block_size = get_block_size(hp_tensor.shape, granularity)
+        block_size = list(block_size)
+
+        # TODO: https://github.com/pytorch/ao/issues/2511
+        # NOTE: fbgemm quant primitives like `torch.ops.fbgemm.quantize_fp8_per_row`
+        # can't pass some of the numerics tests, so we use torchao ones
+        scale = _choose_qparams_affine_float8(
+            hp_tensor, float8_dtype=dtype, block_size=block_size
+        )
+        data = _quantize_affine_float8(hp_tensor, scale, dtype)
+
+        dtype = hp_tensor.dtype
+        del hp_tensor
+        return Float8Tensor(
+            data,
+            scale,
+            block_size=block_size,
+            dtype=dtype,
+        )
+
+
+implements = Float8Tensor.implements
+
+
+@implements([torch.nn.functional.linear, aten.linear.default])
+def _(func, types, args, kwargs):
+    input_tensor, weight_tensor, bias = (
+        args[0],
+        args[1],
+        args[2] if len(args) > 2 else None,
+    )
+    orig_act_size = input_tensor.size()
+    orig_out_features = weight_tensor.shape[-2]
+
+    # NOTE: we are using torchao quant primitive ops instead of fbgemm ones due to various numerical
+    # issues with fbgemm quant primitive ops
+    # for example fbgemm per tensor quant op is not as accurate as the quant primitives in torchao and won't
+    # pass the accuracy check in
+    # TestAffineQuantizedFloat8Compile.test_float8_tensor_slicing_functional_correctness
+    # https://github.com/pytorch/ao/blob/20d45036603c96873b8a8a8391fdbf2a2771ab7e/test/dtypes/test_affine_quantized_float.py#L631
+    # a_data, a_scale = torch.ops.fbgemm.quantize_fp8_per_tensor(input_tensor)
+    if isinstance(input_tensor, Float8Tensor):
+        activation_dtype = input_tensor._data.dtype
+    else:
+        activation_dtype = input_tensor.dtype
+
+    if activation_dtype == torch.float8_e4m3fn:
+        a_data = input_tensor._data
+        a_scale = input_tensor.scale
+        b_data = weight_tensor._data
+        b_scale = weight_tensor.scale.squeeze(-1).contiguous()
+
+        assert a_data.dtype == b_data.dtype
+
+        if _is_per_tensor_quantized(input_tensor):
+            # per tensor quantization
+            res = torch.ops.fbgemm.f8f8bf16(
+                a_data,
+                b_data,
+                a_scale * b_scale,
+                use_fast_accum=True,
+            )
+        else:
+            # per row quantization
```

Comment:
nit: assert for rowwise, to ensure that when someone adds a new scaling type it will be easy to see where support is missing